### DISCUSSION TITLE
What changed when processing subscription batch

### DISCUSSION THREAD

**Discussion by marcuslindblom** [ravendb/ravendb#15664] [Q&A]
Is it possible to see what changes a document has when processing a subscription batch?

Our use case is that we subscribe to the Users collection. If the property `LastLogin` is updated we don't need to run our logic.

**Comment by garayx** [comment]
Hi

You can access the `DocumentStore` of subscription session, inside the subscription batch , so you should be able to wire the ChangeAPI using it.

**Reply by marcuslindblom** [reply]
@garayx do you have an example on how to do that?

**Reply by garayx** [reply]
https://gist.github.com/garayx/51a628f774b6808d2483055323e96414

**Reply by georgiosd** [reply]
If I understand correctly, Marcus wants to know what changed in the document and the subscription was triggered - is that right @marcuslindblom ?

**Reply by marcuslindblom** [reply]
> If I understand correctly, Marcus wants to know what changed in the document and the subscription was triggered - is that right @marcuslindblom ?

That's exactly what I need 👍

**Reply by georgiosd** [reply]
You should be able to adjust your query so that it doesn't get triggered unless something of interest has changed. What is your query like?

**Reply by marcuslindblom** [reply]
Ok, i already do some filtering but I don't see how I can subscribe to Users but if property `lastLogin` was the thing that triggered the subscription I don't want to proceed.

This is the query
```

declare function filter(d) {{
      var wk = load('{ActiveWorkspaceId}');
      return ['Workspaces', 'Teams', 'Users'].includes(d['@metadata']['@collection']) && !d.Deleted && (wk.Users.indexOf(id(d)) != -1
          || wk.Owner == id(d)
          || wk.Teams.indexOf(id(d)) != -1
          || wk.Integrations.indexOf(id(d)) != -1
          || id(d) == '{ActiveWorkspaceId}'
          || d['@metadata']['@collection'] === 'Workspaces' && (d.Users.indexOf('{UserId}') != -1 || d.Owner === '{UserId}'));
      }}
      from @all_docs as doc
      where filter(doc)"```

**Reply by georgiosd** [reply]
So why not add `LastLogin` in this function? If you need to know the previous value, well, you need to store the previous value. e.g. you can store the last x login timestamps.

**Reply by marcuslindblom** [reply]
Ok, there are a couple of more properties that I don't want to trigger my code so I t feels a bit clunky. I was hoping for a built in function for this 😀

**Reply by ayende** [reply]
You can use versioned subscriptions, that will give you the previous & current state of the document, so you can make a determination based on whatever the difference is.

**Reply by ayende** [reply]
See more details here:
https://ravendb.net/learn/inside-ravendb-book/reader/4.0/5-batch-processing-with-subscriptions#versioned-subscriptions

**Reply by marcuslindblom** [reply]
@ayende Ok I see.

I use a projection in my subscription like this.

```
from @all_docs as doc
    select {{
        id: id(doc),
        doc: doc,
        path: getUrl(doc),
        parent: doc.origin.id
    }}";
```

and this is my subscription result class

```c#
  public class ModelsSubscriptionResult
  {
    [JsonProperty("id")]
    public string Id { get; set; }
    [JsonProperty("path")]
    public string Path { get; set; }
    [JsonProperty("parent")]
    public string Parent { get; set; }
    [JsonProperty("doc")]
    public dynamic Doc { get; set; }
  }
```

If I add `from @all_docs (Revisions = true) as doc` I guess I have to return both current and previous in my projection?

Also, this does not seem to work?

![Screenshot 2023-01-24 at 16 21 16](https://user-images.githubusercontent.com/319720/214334415-d4b04f8c-9210-40e5-9f09-cdec271e34da.png)

![Screenshot 2023-01-24 at 16 37 43](https://user-images.githubusercontent.com/319720/214338277-98e8bdcf-58ee-4473-bbf9-9a14df43b1f2.png)

**Reply by garayx** [reply]
you can return them both, depend on you worker logic, yes.

What you mean by, it doesn't seems to work?

**Reply by marcuslindblom** [reply]
@garayx Well the value is `null` for both `Current` and `Previous`, should be a document?

**Reply by garayx** [reply]
do you actually have a revision for files/1-A ? if yes how many ? 

IIRC both Current and Previous are null if the document was deleted, is it the case?

**Reply by ayende** [reply]
Tested this:
![image](https://user-images.githubusercontent.com/116915/214607332-0b4a8cdc-0280-4f1c-869e-9f8eac45e875.png)


And this:

![image](https://user-images.githubusercontent.com/116915/214607396-9ca50800-5af0-4e61-833f-27bf3d4e79c0.png)


They both appear to work as expected

**Reply by marcuslindblom** [reply]
I had somewhat over 60k documents with the id `temps/...` after removing all revisions it seems to work better. Why did I have so many deleted revisons with id `temps/...`?

**Reply by garayx** [reply]
hi,

you can see here how the revisions are working:
https://ravendb.net/docs/article-page/5.4/csharp/document-extensions/revisions/overview#how-it-works
